src/SharpBrick.PoweredUp.BlueZ/BlueZPoweredUpBluetoothDevice.cs

```diff
@@ -0,0 +1,136 @@
+using Tmds.DBus;
+
+namespace SharpBrick.PoweredUp.BlueZ
+{
+    internal class BlueZPoweredUpBluetoothDevice : IPoweredUpBluetoothDevice
+    {
+        private Func<PoweredUpBluetoothDeviceInfo, Task> _discoveryHandler;
+        private IDevice1 _device;
+
+        internal PoweredUpBluetoothDeviceInfo DeviceInfo { get; private set; } = new PoweredUpBluetoothDeviceInfo();
+        internal bool Connected { get; private set; } = false;
+        internal bool ServicesResolved { get; private set;} = false;
+
+        internal BlueZPoweredUpBluetoothDevice(IDevice1 device, Func<PoweredUpBluetoothDeviceInfo, Task> discoveryHandler = null)
+        {
+            _discoveryHandler = discoveryHandler;
+            _device = device;
+        }
+
+        public string Name { get; private set; } = string.Empty;
+
+        private async Task InvokeDiscoveryHandlerAsync()
+        {
+            if (_discoveryHandler != null && DeviceInfo?.ManufacturerData != null)
+            {
+                var tempHandle = _discoveryHandler;
+                _discoveryHandler = null; // make sure we only execute the handler exactly once.
+                await tempHandle(DeviceInfo);
+            }
+        }
+
+        internal async Task Initialize()
+        {
+            await _device.WatchPropertiesAsync(DevicePropertyChangedHandler);
+
+            await GetSafeDeviceInfoAsync();
+        }
+
+        internal async Task TryGetManufacturerDataAsync()
+        {
+            try 
+            {
+                var manufacturerData = await _device.GetManufacturerDataAsync();
+                DeviceInfo.ManufacturerData = (byte[])manufacturerData.First().Value;
+            } 
+            catch 
+            {
+                // we can ignore errors here, this will throw an exception for existing devices from a previous session (only after reboot/restart of bluetoothd)
+                // for these device, manufacturer data will be returned when discovery is turned on
+            }
+        }
+
+        private async void DevicePropertyChangedHandler(PropertyChanges changes)
+        {
+            foreach (var propertyChanged in changes.Changed)
+            {
+                switch (propertyChanged.Key)
+                {
+                    case "ManufacturerData":
+                        DeviceInfo.ManufacturerData = (byte[])((IDictionary<ushort,object>)propertyChanged.Value).First().Value;
+                        break;
+                    case "Connected":
+                        Connected = (bool)propertyChanged.Value;
+                        break;
+                    case "ServicesResolved":
+                        ServicesResolved = (bool)propertyChanged.Value;
+                        break;
+                    case "RSSI":
+                        // this is the only dependable property that will change from a pre-existing device during discovery
+                        await InvokeDiscoveryHandlerAsync();
```

Comment:
Yeah the difficulty is handling completely new **and** previously connected devices using bluez/dbus.

BlueZ will notify new devices (and services for that matter) being added, for example during discovery.
But for existing devices, this will never happen (unless you remove the device first). 
Then, existing devices come in two flavors: ones that you have connected to since the last bluetooth daemon restart and ones that you connected to before that. 

But invoking the discovery handler for existing devices is only supposed to happen when you turn on discovery and you actually turn on the device, not immediately, otherwise you would 'find' any hubs you have ever connected to directly and for example, most of the examples will try to connect to this hub immediately, when it might still be off. 

So I figured I could rely on the manufacturer data, since that **is** sent during discovery, for existing devices. But the property changed event will only fire if it changes, so for existing devices that were connected during the same bluetooth daemon session (say, since last reboot) the manufacturer data will be there already. Therefore no change, not event, no discovery handler. 
So the only thing dependable I found was to use the RSSI property. Now this might introduce some delays before connecting in some cases, but it seems to work fine.

So, long story short, yes this is far from pretty, but it's the only way I managed to make it work in all cases. And the root cause is the way bluez announces things using dbus - and that is by design apparently 😐